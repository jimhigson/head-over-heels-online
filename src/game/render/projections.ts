import type { Container } from "pixi.js";
import { blockSizePx } from "../../sprites/spritePivots";
import {
  type XyMaybeZ,
  type Xyz,
  type Xy,
  type OrthoPlane,
  addXyz,
  subXy,
} from "../../utils/vectors/vectors";

export const moveContainerToBlockXyz = (
  blockXyz: XyMaybeZ,
  sprite: Container,
) => {
  const projectionXyz = projectBlockXyzToScreenXy(blockXyz);

  sprite.x = projectionXyz.x;
  sprite.y = projectionXyz.y;

  return sprite;
};

/* position on 2d screen for a given xyz in game-space 3d pixels */
export const projectWorldXyzToScreenX = ({
  x = 0,
  y = 0,
}: Partial<Xyz>): number => y - x;

export const projectWorldXyzToScreenXy = ({
  x: xw = 0,
  y: yw = 0,
  z: zw = 0,
}: Partial<Xyz>): Xy => {
  return {
    x: yw - xw,
    y: -(xw + yw) / 2 - zw,
  };
};

/**
 * for the editor - convert screen (mouse) xy to world xyz, on
 * the visible surface of an object
 */
export const unprojectScreenXyToWorldXyzOnFace = (
  /** the screen x,y in game engine (scaled) pixels */
  scrPos: Xy,
  itemPosition: Xyz,
  plane: OrthoPlane,
): Xyz => {
  const itemOriginScreen = projectWorldXyzToScreenXy(itemPosition);

  // adjust xs, xy to be relative to the projected-on-screen origin of the item:
  const scrPosAdj = subXy(scrPos, itemOriginScreen);

  return addXyz(unprojectScreenXyToWorldXyz(scrPosAdj, plane), itemPosition);
};

/**
 * for the editor - convert screen (mouse) xy to world xyz, for a plane
 * starting at the origin
 */
export const unprojectScreenXyToWorldXyz = (
  /** the screen x,y in game engine (scaled) pixels */
  { x: xScr, y: yScr }: Xy,
  plane: OrthoPlane,
): Xyz => {
  // adjust xs, xy to be relative to the projected-on-screen origin of the item:

  switch (plane) {
    case "xy": {
      // top/bottom face
      return {
        x: -(xScr / 2 + yScr),
        y: xScr / 2 - yScr,
        z: 0,
      };
    }
    case "xz": {
      // away/towards face
      return {
        x: -xScr,
        y: 0,
        z: xScr / 2 - yScr,
      };
    }
    case "yz": {
      // left/right face
      return {
        x: 0,
        y: xScr,
        z: -xScr / 2 - yScr,
      };
    }
    default:
      plane satisfies never;
      throw new Error(`unknown plane ${plane}`);
  }
};

/** get the in-game x,y,z for any given block x,y,z */
export const blockXyzToFineXyz = ({
  x: xb = 0,
  y: yb = 0,
  z: zb = 0,
}: Partial<Xyz>): Xyz => {
  return {
    x: xb * blockSizePx.w,
    y: yb * blockSizePx.d,
    z: zb * blockSizePx.h,
  };
};

/** get the in-game x,y,z for any given block x,y,z */
export const fineXyzToBlockXyz = ({
  x: xf = 0,
  y: yf = 0,
  z: zf = 0,
}: Partial<Xyz>): Xyz => {
  return {
    x: xf / blockSizePx.w,
    y: yf / blockSizePx.d,
    z: zf / blockSizePx.h,
  };
};

/* position on 2d screen for a given xyz in game-space block position */
export const projectBlockXyzToScreenXy = (blockXyz: Partial<Xyz>): Xy => {
  return projectWorldXyzToScreenXy(blockXyzToFineXyz(blockXyz));
};
