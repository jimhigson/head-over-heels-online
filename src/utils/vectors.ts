export const directions = ["away", "towards", "left", "right"] as const;
export type Direction = (typeof directions)[number];

export const directionVectors: Record<Direction, Xyz> = {
  away: { x: 0, y: 1, z: 0 },
  left: { x: 1, y: 0, z: 0 },
  right: { x: -1, y: 0, z: 0 },
  towards: { x: 0, y: -1, z: 0 },
};

export type Xy = {
  x: number;
  y: number;
};

export const crossAxisXy = (axis: AxisXy): AxisXy => (axis === "x" ? "y" : "x");

export const addXy = (xy: Xy, ...xys: Array<Partial<Xy>>): Xy =>
  xys.reduce<Xy>(
    (ac, xyi) => ({
      x: ac.x + (xyi.x ?? 0),
      y: ac.y + (xyi.y ?? 0),
    }),
    xy,
  );

export const scaleXyz = (xy: Xyz, scale: number): Xyz => ({
  x: xy.x * scale,
  y: xy.y * scale,
  z: xy.z * scale,
});

export const addXyz = (xyz: Xyz, ...xyzs: Array<Partial<Xyz>>): Xyz =>
  xyzs.reduce<Xyz>(
    (ac, xyi) => ({
      x: ac.x + (xyi.x ?? 0),
      y: ac.y + (xyi.y ?? 0),
      z: ac.z + (xyi.z ?? 0),
    }),
    xyz,
  );

export const xyzEqual = (
  { x: ax, y: ay, z: az }: Xyz,
  { x: bx, y: by, z: bz }: Xyz,
) => {
  return ax === bx && ay === by && az === bz;
};

export const originXyz: Xyz = { x: 0, y: 0, z: 0 };

export type Xyz = {
  x: number;
  y: number;
  z: number;
};
export type XyMaybeZ = {
  x: number;
  y: number;
  z?: number;
};
export type AxisXy = "x" | "y";

export const axesXyz = ["x", "y", "z"] as const;
export type AxisXyz = (typeof axesXyz)[number];

//export type Aabb = [Xyz, Xyz];
/** how big is the collision box around an item? */
export type Aabb = Xyz;
