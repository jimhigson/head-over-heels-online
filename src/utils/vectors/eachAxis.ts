import type { Xyz } from "./vectors";

/**
 * Convenience for setting each of x,y,z or an Xyz vector based on zero or more
 * vectors
 */
export const eachAxis = (fn: (...n: number[]) => number, ...vectors: Xyz[]) => {
  return {
    x: fn(...vectors.map((p) => p.x)),
    y: fn(...vectors.map((p) => p.y)),
    z: fn(...vectors.map((p) => p.z)),
  };
};
