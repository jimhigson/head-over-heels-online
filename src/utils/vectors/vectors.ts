export const directions4Xy = ["away", "towards", "left", "right"] as const;
export type Direction4Xy = (typeof directions4Xy)[number];

export const directionsXyDiagonal = [
  "awayRight",
  "rightTowards",
  "towardsLeft",
  "leftAway",
] as const;

export type DirectionXyDiagonal = (typeof directionsXyDiagonal)[number];

export const directions8Xy = [
  ...directions4Xy,
  ...directionsXyDiagonal,
] as const;
export type Direction8Xy = (typeof directions8Xy)[number];

// prettier-ignore
type Matrix3x3 = [
  number, number, number, // Row 1
  number, number, number, // Row 2
  number, number, number, // Row 3
];

export const directionAxis = (direction: Direction4Xy): AxisXy =>
  direction === "away" || direction === "towards" ? "y" : "x";

export const oppositeDirection = (direction: Direction4Xy): Direction4Xy =>
  direction === "away" ? "towards"
  : direction === "towards" ? "away"
  : direction === "left" ? "right"
  : "left";

/**
 * doors sit along the axis perpendicular to their direction
 */
export const doorAlongAxis = (doorDirection: Direction4Xy): AxisXy =>
  perpendicularAxisXy(directionAxis(doorDirection));

export type DirectionZ = "down" | "up";
export type Direction4Xyz = Direction4Xy | DirectionZ;
export type Direction8Xyz = Direction8Xy | DirectionZ;

export type Xy = {
  x: number;
  y: number;
};

export const perpendicularAxisXy = (axis: AxisXy): AxisXy =>
  axis === "x" ? "y" : "x";

export const addXy = (xy: Xy, ...xys: Array<Partial<Xy>>): Xy =>
  xys.reduce<Xy>(
    (ac, xyi) => ({
      x: ac.x + (xyi.x ?? 0),
      y: ac.y + (xyi.y ?? 0),
    }),
    xy,
  );

export const subXy = (xy: Xy, ...xys: Array<Partial<Xy>>): Xy =>
  xys.reduce<Xy>(
    (ac, xyi) => ({
      x: ac.x - (xyi.x ?? 0),
      y: ac.y - (xyi.y ?? 0),
    }),
    xy,
  );

export const scaleXyz = (xy: Xyz, scale: number): Xyz => ({
  x: xy.x * scale,
  y: xy.y * scale,
  z: xy.z * scale,
});

export const xyzMagnitude = ({ x, y, z }: Xyz) =>
  Math.sqrt(x * x + y * y + z * z);

export const unitVector = (xyz: Xyz): Xyz =>
  scaleXyz(xyz, 1 / xyzMagnitude(xyz));

/**
 * clockwise rotation of the xy component
 */
export const perpendicularXyz = ({ x, y, z }: Xyz): Xyz => ({
  x: y,
  y: -x,
  z,
});

export const addXyz = (...xyzs: Array<Partial<Xyz>>): Xyz => {
  return xyzs.reduce<Xyz>(
    ({ x: acX, y: acY, z: acZ }, { x: iX = 0, y: iY = 0, z: iZ = 0 }) => ({
      x: acX + iX,
      y: acY + iY,
      z: acZ + iZ,
    }),
    originXyz,
  );
};

export const subXyz = (xyz: Xyz, ...xyzs: Array<Partial<Xyz>>): Xyz =>
  xyzs.reduce<Xyz>(
    (ac, xyi) => ({
      x: ac.x - (xyi.x ?? 0),
      y: ac.y - (xyi.y ?? 0),
      z: ac.z - (xyi.z ?? 0),
    }),
    xyz,
  );

export const xyzEqual = (
  { x: ax, y: ay, z: az }: Xyz,
  { x: bx, y: by, z: bz }: Xyz,
) => {
  return ax === bx && ay === by && az === bz;
};
export const xyEqual = ({ x: ax, y: ay }: Xy, { x: bx, y: by }: Xy) => {
  return ax === bx && ay === by;
};

/**
 * because of floating point error, after processing mtv it is possible to get
 * values that should be integers but are off by a tiny amount. To correct, we consider
 * anything that is within 1/1000 of a pixel to be exactly on that pixel
 */
const isIntegerOrCloseTo = (n: number) => Math.abs(n - Math.round(n)) < 0.001;
/* is the xyz in integer position (or very close to) */
export const isIntegerXyzOrCloseTo = ({ x = 0, y = 0, z = 0 }: Partial<Xyz>) =>
  isIntegerOrCloseTo(x) && isIntegerOrCloseTo(y) && isIntegerOrCloseTo(z);

export const isExactIntegerXyz = ({ x = 0, y = 0, z = 0 }: Partial<Xyz>) =>
  Number.isInteger(x) && Number.isInteger(y) && Number.isInteger(z);

export const roundXyz = ({ x, y, z }: Xyz) => ({
  x: Math.round(x),
  y: Math.round(y),
  z: Math.round(z),
});

export const originXy: Xy = Object.freeze({ x: 0, y: 0 });
export const originXyz: Xyz = Object.freeze({ x: 0, y: 0, z: 0 });

export type Xyz = {
  x: number;
  y: number;
  z: number;
};
export type XyMaybeZ = {
  x: number;
  y: number;
  z?: number;
};
export type AxisXy = "x" | "y";

export const axesXyz = ["x", "y", "z"] as const;
export type AxisXyz = (typeof axesXyz)[number];

export type Aabb = Readonly<Xyz>;

/**
 * dot product - the component of one vector in the direction of another
 */
export const dotProductXyz = (a: Xyz, b: Xyz): number => {
  return a.x * b.x + a.y * b.y + a.z * b.z;
};

export const multiplyMatrixVector = (
  matrix: Matrix3x3,
  { x, y, z }: Xyz,
): Xyz => ({
  x: matrix[0] * x + matrix[1] * y + matrix[2] * z,
  y: matrix[3] * x + matrix[4] * y + matrix[5] * z,
  z: matrix[6] * x + matrix[7] * y + matrix[8] * z,
});

// prettier-ignore
/** matrix to multiply against to discard the z component of a vector */
export const xyOnlyMatrix : Matrix3x3 = [
  1, 0, 0, // Row 1
  0, 1, 0, // Row 2
  0, 0, 0, // Row 3
];

export const vectorClosestDirectionXy4 = ({ x, y }: Xy): Direction4Xy => {
  if (y > x) {
    if (y > -x) {
      return "away";
    } else {
      return "right";
    }
  } else {
    if (y > -x) {
      return "left";
    } else {
      return "towards";
    }
  }
};

export const distanceXySquared = (
  { x: x1, y: y1 }: Xyz,
  { x: x2, y: y2 }: Xyz,
) => {
  return Math.abs(x2 - x1) ** 2 + Math.abs(y2 - y1) ** 2;
};
