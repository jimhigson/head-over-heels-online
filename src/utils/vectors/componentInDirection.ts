import { dotProductXyz, scaleXyz, type Xyz } from "./vectors";

/**
 * returns the component of vector @param v in the direction described by unit
 * vector @param unitD */

export const componentInDirection = (v: Xyz, unitD: Xyz): Xyz => {
  // Calculate the scalar projection of v onto d
  const scalarProjection = dotProductXyz(v, unitD);

  // Scale the unit vector d by the scalar projection to get the vector component
  return scaleXyz(unitD, scalarProjection);
};
