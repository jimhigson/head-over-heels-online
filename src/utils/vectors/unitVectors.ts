import type { Direction8Xyz, Xyz } from "./vectors";

import { addXyz, unitVector } from "./vectors";

const l = { x: 1, y: 0, z: 0 };
const r = { x: -1, y: 0, z: 0 };
const t = { x: 0, y: -1, z: 0 };
const a = { x: 0, y: 1, z: 0 };

export const unitVectors: Record<Direction8Xyz, Xyz> = {
  away: a,
  left: l,
  right: r,
  towards: t,
  down: { x: 0, y: 0, z: -1 },
  up: { x: 0, y: 0, z: 1 },
  awayRight: unitVector(addXyz(a, r)),
  towardsRight: unitVector(addXyz(t, r)),
  towardsLeft: unitVector(addXyz(t, l)),
  awayLeft: unitVector(addXyz(a, l)),
};
