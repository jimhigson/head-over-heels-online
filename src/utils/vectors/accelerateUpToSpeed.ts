import {
  type Xyz,
  scaleXyz,
  dotProductXyz,
  addXyz,
  subXyz,
  perpendicularXyz,
  xyzLength,
} from "@/utils/vectors/vectors";
import { componentInDirection } from "./componentInDirection";

/** accelerates the current velocity towards the given max speed,
 * reaching it if close enough rather than allowing to go over
 * @return the new velocity, after acceleration has been applied
 */
export const accelerateToSpeed = ({
  vel,
  acc,
  unitD,
  maxSpeed,
  deltaMS,
  // how much to fade the cross-component of the acellerated direction.
  // eg, if you're accelerating north, how much to slow down going east?
  crossComponentFade = 0,
  minVelocity = 0,
}: {
  /* The current velocity to accelerate from */
  vel: Xyz;
  /**
   * acceleration to apply, in m/s²
   */
  acc: number;

  crossComponentFade?: number;

  minVelocity?: number;

  /** unit vector of the direction we want to accelerate in */
  unitD: Xyz;
  maxSpeed: number;
  deltaMS: number;
}): Xyz => {
  const existingSpeedInDirection = dotProductXyz(vel, unitD);

  const targetSpeedInDirection = Math.max(
    existingSpeedInDirection + acc * deltaMS,
    minVelocity,
  );

  // would applying the full accel put us over the max speed?
  const hitsMaxSpeedInDirection = targetSpeedInDirection >= maxSpeed;

  const crossUnitD = perpendicularXyz(unitD);
  const exitingSpeedInCrossDirection = dotProductXyz(vel, crossUnitD);
  const newSpeedInCrossDirection =
    exitingSpeedInCrossDirection >= 0 ?
      Math.max(exitingSpeedInCrossDirection - crossComponentFade * deltaMS, 0)
    : Math.min(exitingSpeedInCrossDirection + crossComponentFade * deltaMS, 0);

  return addXyz(
    scaleXyz(
      unitD,
      hitsMaxSpeedInDirection ? maxSpeed : targetSpeedInDirection,
    ),
    scaleXyz(crossUnitD, newSpeedInCrossDirection),
  );
};

/**
 * A simpler acceleration that preserves all of the built-up speed, even if it is in a different
 * direction (with no dot=product with the target direction)
 */
export const accelerateToSpeed2 = ({
  vel,
  acc,
  unitD,
  maxSpeed,
  deltaMS,
  minSpeed = 0,
}: {
  /* The current velocity to accelerate from */
  vel: Xyz;
  /**
   * acceleration to apply, in m/s²
   */
  acc: number;

  minSpeed?: number;

  /** unit vector of the direction we want to accelerate in */
  unitD: Xyz;
  maxSpeed: number;
  deltaMS: number;
}): Xyz => {
  const existingSpeed = xyzLength(vel);

  const targetSpeedInDirection = Math.max(
    minSpeed,
    Math.min(maxSpeed, existingSpeed + acc * deltaMS),
  );

  // would applying the full accel put us over the max speed?
  const newSpeed = Math.min(targetSpeedInDirection, maxSpeed);

  return scaleXyz(unitD, newSpeed);
};

/* calculated the accel vector to exactly hit a speed right away, in this frame */
export const instantAccelToSpeed = ({
  vel,
  unitD,
  speed,
  deltaMS,
}: {
  /* The current velocity to accelerate from */ vel: Xyz;
  unitD: Xyz /* speed to hit, in px/ms */;
  speed: number /* how long to accel over */;
  deltaMS: number;
}) => {
  return scaleXyz(
    subXyz(scaleXyz(unitD, speed), componentInDirection(vel, unitD)),
    1 / deltaMS,
  );
};
