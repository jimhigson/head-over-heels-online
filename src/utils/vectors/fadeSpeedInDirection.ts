import { componentInDirection } from "./componentInDirection";
import type { Xyz } from "./vectors";
import {
  xyzEqual,
  originXyz,
  unitVector,
  scaleXyz,
  dotProductXyz,
  addXyz,
} from "./vectors";

/*
 * eg, for when a direction is no longer pressed and player is idle - basically, friction etc
 */
export const fadeSpeedInDirection = ({
  vel,
  travelDirection,
  deceleration,
  deltaMS,
}: {
  vel: Xyz /**
   * The direction currently going in, to slow down from (or the component of travel).
   * The magnitude is not important since this function makes a unit vector from this vector
   */;
  travelDirection: Xyz;
  deceleration: number;
  deltaMS: number;
}): Xyz => {
  if (xyzEqual(travelDirection, originXyz)) {
    return originXyz;
  }

  const travelDirectionUnit = unitVector(travelDirection);
  const decelerationVector = scaleXyz(travelDirectionUnit, deceleration);

  const velocityComponentInDirection = componentInDirection(
    vel,
    travelDirectionUnit,
  );

  return (
      dotProductXyz(
        travelDirectionUnit,
        addXyz(vel, scaleXyz(decelerationVector, deltaMS)),
      ) < 0
    ) ?
      // applying full deceleration would go past zero and start moving in the opposite direction
      // vel here is wrong - should be the component of vel in the direction of deceleration!
      scaleXyz(velocityComponentInDirection, -1 / deltaMS)
      // maximum deceleration wouldn't take past zero so return that
    : decelerationVector;
};
