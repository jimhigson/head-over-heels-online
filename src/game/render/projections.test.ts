import { describe, expect, test } from "vitest";
import type { OrthoPlane, Xy, Xyz } from "../../utils/vectors/vectors";
import { originXyz } from "../../utils/vectors/vectors";
import { unprojectScreenXyToWorldXyzOnFace } from "./projections";

const smallCubeAabb: Xyz = {
  x: 10,
  y: 10,
  z: 10,
};

describe("unprojectScreenXyToWorldXyzOnFace", () => {
  type TestParams = {
    desc: string;
    screen: Xy;
    itemPosition: Xyz;
    itemAabb: Xyz;
    face: OrthoPlane;
    expected: Xyz;
  };

  const runTest = ({
    face,
    itemPosition,
    itemAabb,
    screen,
    expected,
  }: TestParams) => {
    expect(
      unprojectScreenXyToWorldXyzOnFace(screen, itemPosition, itemAabb, face),
    ).toMatchObject({
      x: expect.closeTo(expected.x, 6),
      y: expect.closeTo(expected.y, 6),
      z: expect.closeTo(expected.z, 6),
    });
  };
  const testDesc =
    "$desc : unprojectScreenXyToWorldXyzOnFace(screen=$screen, position=$itemPosition, aabb=$itemAabb, face=$face) = $expected";

  describe("xy(top) face", () => {
    describe("item at origin", () => {
      test.each<TestParams>([
        {
          desc: "front corner",
          screen: { x: 0, y: -10 },
          itemPosition: originXyz,
          itemAabb: smallCubeAabb,
          face: "xy",
          expected: { x: 0, y: 0, z: 10 },
        },
        {
          desc: "back corner",
          screen: { x: 0, y: -20 },
          itemPosition: originXyz,
          itemAabb: smallCubeAabb,
          face: "xy",
          expected: { x: 10, y: 10, z: 10 },
        },
        {
          desc: "visually left corner",
          screen: { x: -10, y: -15 },
          itemPosition: originXyz,
          itemAabb: smallCubeAabb,
          face: "xy",
          expected: { x: 10, y: 0, z: 10 },
        },
      ])(testDesc, runTest);
    });
  });
});
