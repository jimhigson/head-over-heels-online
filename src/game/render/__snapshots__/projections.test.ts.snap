// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unprojectScreenXyToWorldXyzOnFace > xy(top) face > item at origin > 'back corner' : unprojectScreenXyToWorldXyzOnFace(screen={ x: +0, y: -20 }, position={ x: +0, y: +0, z: +0 }, aabb=undefined, face='xy') = { x: 10, y: 10, z: 10 } 1`] = `
{
  "x": 20,
  "y": 20,
  "z": 0,
}
`;

exports[`unprojectScreenXyToWorldXyzOnFace > xy(top) face > item at origin > 'back corner' : unprojectScreenXyToWorldXyzOnFace(screen={ x: +0, y: -20 }, position={ x: +0, y: +0, z: +0 }, aabb=undefined, face='xy') = undefined 1`] = `
{
  "x": 20,
  "y": 20,
  "z": 0,
}
`;

exports[`unprojectScreenXyToWorldXyzOnFace > xy(top) face > item at origin > 'front corner' : unprojectScreenXyToWorldXyzOnFace(screen={ x: +0, y: -10 }, position={ x: +0, y: +0, z: +0 }, aabb=undefined, face='xy') = { x: +0, y: +0, z: 10 } 1`] = `
{
  "x": 10,
  "y": 10,
  "z": 0,
}
`;

exports[`unprojectScreenXyToWorldXyzOnFace > xy(top) face > item at origin > 'front corner' : unprojectScreenXyToWorldXyzOnFace(screen={ x: +0, y: -10 }, position={ x: +0, y: +0, z: +0 }, aabb=undefined, face='xy') = undefined 1`] = `
{
  "x": 10,
  "y": 10,
  "z": 0,
}
`;

exports[`unprojectScreenXyToWorldXyzOnFace > xy(top) face > item at origin > 'visually left corner' : unprojectScreenXyToWorldXyzOnFace(screen={ x: -10, y: -15 }, position={ x: +0, y: +0, z: +0 }, aabb=undefined, face='xy') = { x: 10, y: +0, z: 10 } 1`] = `
{
  "x": 20,
  "y": 10,
  "z": 0,
}
`;

exports[`unprojectScreenXyToWorldXyzOnFace > xy(top) face > item at origin > 'visually left corner' : unprojectScreenXyToWorldXyzOnFace(screen={ x: -10, y: -15 }, position={ x: +0, y: +0, z: +0 }, aabb=undefined, face='xy') = undefined 1`] = `
{
  "x": 20,
  "y": 10,
  "z": 0,
}
`;
